### ISSUE TITLE
Physics2D: Multi-Threading

### ISSUE THREAD

**Issue by ret80** [godotengine/godot#6512]
**Operating system or device - Godot version:**
Godot 2.1 Win

**Issue description** (what happened, and what was expected):
I turned on "Multi-Threading" for physics. And often after a while an error appears:

```
ERROR: Map<struct BroadPhase2DHashGrid::Element *,struct BroadPhase2DHashGrid::P
airData *,struct Comparator<struct BroadPhase2DHashGrid::Element *>,class Defaul
tAllocator>::_set_color: Condition ' p_node == _data._nil && p_color == RED ' is
 true.
   At: core\map.h:
```

The application in this case hangs.

**Steps to reproduce:**

**Link to minimal example project** (optional but very welcome):
https://dl.dropboxusercontent.com/u/80689824/PhyTest_MT.zip

**Comment by supaiku-o** [comment]
I experience a similiar issue but for me it happens at:
```GDScript
ERROR: _set_color: Condition ' p_node == _data._nil && p_color == RED ' is true.
   At: core/set.h:152.
```
It appears to always happen somewhere where I make call_deferred calls from a worker thread to initialize and instantiate packed scenes. But it only happens 1 out of a 100 times and I have no clue how to reproduce.

But of course it could be something entirely else, but that's the only place where I noticed it. 

Should I create a separate issue for it?

**Comment by jtaart** [comment]
I experienced the same thing as supaiku-o. It seems very random. I tried to make a test file, but I could only get it to do it once or twice. It does seem to have to do with call_defferred calls in a Thread.

**Comment by raymoo** [comment]
I thought multithreaded physics was still WIP? Crashing would still be a bug, though.

**Comment by raymoo** [comment]
Looks like a race condition in set?

**Comment by jtaart** [comment]
@raymoo: I guess it is a WIP, but still would love to be able to use it. 

I made a test file that loads a ton of Nodes in a thread and queue_frees them. On my pc it causes it to loop error about 50% of the time. Don't think it has anything to do with physics since I used no physics nodes.

Test file: [http://s000.tinyupload.com/?file_id=52870569225830189698](url)

https://i.imgur.com/xO8hl8P.gif

**Comment by raymoo** [comment]
Maybe some race condition with the call_deferred queue then? I haven't looked at all into this so I am just guessing.

**Comment by jtaart** [comment]
@supaiku-o  In the test file I only have one call_deferred. its used to  call the function with wait_to_finish() later. I no longer think its has something to do with call_deferred.

**Comment by jtaart** [comment]
Nvm I was using 2.1.3. It doesn't seem to do it in 2.1.4. This might be fixed.

**Comment by raymoo** [comment]
@jtaart There can still be a race condition if multiple threads attempt to access the call queue at the same time.

**Comment by Noshyaar** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about it here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*

**Comment by akien-mga** [comment]
As there was no answer to the above request, we're closing this issue as inactive.
If you can still reproduce it, please open a new issue against a currently supported release of Godot.
